src/cpp/ext/filters/census/environment_autodetect.cc

```diff
@@ -0,0 +1,337 @@
+}
+
+// Get module ID for App Engine
+std::string GetModuleId() {
+  return grpc_core::GetEnv("GAE_SERVICE").value_or("");
+}
+
+// Get version ID for App Engine
+std::string GetVersionId() {
+  return grpc_core::GetEnv("GAE_VERSION").value_or("");
+}
+
+// Fire and forget class
+class EnvironmentAutoDetectHelper {
+ public:
+  explicit EnvironmentAutoDetectHelper(
+      std::string project_id, grpc_polling_entity* pollent,
+      absl::AnyInvocable<void(EnvironmentAutoDetect::ResourceType)> on_done)
+      : project_id_(std::move(project_id)),
+        pollent_(pollent),
+        on_done_(std::move(on_done)) {
+    grpc_core::MutexLock lock(&mu_);
+    // GKE
+    resource_.labels.emplace("project_id", project_id_);
+    if (grpc_core::GetEnv("KUBERNETES_SERVICE_HOST").has_value()) {
+      resource_.resource_type = "k8s_container";
+      resource_.labels.emplace("namespace_name", GetNamespaceName());
+      resource_.labels.emplace("pod_name", GetPodName());
+      resource_.labels.emplace("container_name", GetContainerName());
+      attributes_to_fetch_.push_back({"location", kZoneAttribute});
+      attributes_to_fetch_.push_back({"cluster_name", kClusterNameAttribute});
+    }
+    // Cloud Functions
+    else if (grpc_core::GetEnv("FUNCTION_NAME").has_value() ||
+             grpc_core::GetEnv("FUNCTION_TARGET").has_value()) {
+      resource_.resource_type = "cloud_function";
+      resource_.labels.emplace("function_name", GetFunctionName());
+      attributes_to_fetch_.push_back({"region", kRegionAttribute});
+    }
+    // Cloud Run
+    else if (grpc_core::GetEnv("K_CONFIGURATION").has_value()) {
+      resource_.resource_type = "cloud_run_revision";
+      resource_.labels.emplace("revision_name", GetRevisionName());
+      resource_.labels.emplace("service_name", GetServiceName());
+      resource_.labels.emplace("configuration_name", GetConfiguratioName());
+      attributes_to_fetch_.push_back({"location", kRegionAttribute});
+    }
+    // App Engine
+    else if (grpc_core::GetEnv("GAE_SERVICE").has_value()) {
+      resource_.resource_type = "gae_app";
+      resource_.labels.emplace("module_id", GetModuleId());
+      resource_.labels.emplace("version_id", GetVersionId());
+      attributes_to_fetch_.push_back({"zone", kZoneAttribute});
+    }
+    // Assume GCE
+    else {
+      assuming_gce_ = true;
+      resource_.resource_type = "gce_instance";
+      attributes_to_fetch_.push_back({"instance_id", kInstanceIdAttribute});
+      attributes_to_fetch_.push_back({"zone", kZoneAttribute});
+    }
+    FetchMetadataServerAttributesAsynchronouslyLocked();
+  }
+
+ private:
+  struct Attribute {
+    std::string resource_attribute;
+    std::string metadata_server_atttribute;
+  };
+
+  const std::string project_id_;
+  grpc_polling_entity* pollent_;
+  absl::AnyInvocable<void(EnvironmentAutoDetect::ResourceType)> on_done_;
+  grpc_core::Mutex mu_;
+  std::list<Attribute> attributes_to_fetch_ ABSL_GUARDED_BY(mu_);
```

Comment:
This seems like it would be simpler & clearer if it was a `flat_hash_map<string /* metadata_server_attribute */, string /* resource_attribute */>`

Reply:
suggestion taken